Sources/LocheckLogic/Problems.swift

```diff
@@ -0,0 +1,264 @@
+//
+//  Problems.swift
+//
+//
+//  Created by Steve Landey on 8/27/21.
+//
+
+import Foundation
+
+protocol StringsdictProblem: SummarizableProblem {
+    var key: String { get }
+}
+
+protocol StringsProblem: SummarizableProblem {
+    var key: String { get }
+}
+
+struct LprojFileMissingFromTranslation: Problem, Equatable {
+    var kindIdentifier: String { "lproj_file_missing_from_translation" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .warning }
+    let key: String
+    let language: String
+
+    var message: String { "\(key) missing from \(language)" }
+}
+
+struct StringHasDuplicateArguments: Problem, StringsProblem, Equatable {
+    var kindIdentifier: String { "string_has_duplicate_arguments" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .warning }
+    let key: String
+    let language: String
+
+    var message: String {
+        "Some arguments appear more than once in this translation"
+    }
+}
+
+struct StringHasExtraArguments: Problem, StringsProblem, Equatable {
+    var kindIdentifier: String { "string_has_extra_arguments" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .warning }
+    let key: String
+    let language: String
+    let args: [String]
+
+    var message: String {
+        "Translation includes arguments that don't exist in the source: \(args.joined(separator: ", "))"
+    }
+}
+
+struct StringHasInvalidArgument: Problem, StringsProblem, Equatable {
+    var kindIdentifier: String { "string_has_invalid_argument" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .error }
+    let key: String
+    let language: String
+    let argPosition: Int
+    let baseArgSpecifier: String
+    let argSpecifier: String
+
+    var message: String {
+        "Specifier for argument \(argPosition) does not match (should be \(baseArgSpecifier), is \(argSpecifier))"
+    }
+}
+
+struct StringHasMissingArguments: Problem, StringsProblem, Equatable {
+    var kindIdentifier: String { "string_has_missing_arguments" }
+    var uniquifyingInformation: String { "\(language)-\(key)" }
+    var severity: Severity { .warning }
+    let key: String
+    let language: String
+    let args: [String]
+
+    var message: String { "Does not include argument(s) at \(args.joined(separator: ", "))" }
```

Comment:
[nit] Consider checking `args.count` to avoid `argument(s)`

Reply:
I thought about this but decided to keep all the logic non-branching for generating error messages.